core/src/main/java/org/apache/calcite/rel/core/Sort.java

```diff
@@ -124,11 +125,19 @@ public abstract Sort copy(RelTraitSet traitSet, RelNode newInput,
 
   @Override public @Nullable RelOptCost computeSelfCost(RelOptPlanner planner,
       RelMetadataQuery mq) {
+    double rowCount = mq.getRowCount(this);
+    if (collation.getFieldCollations().isEmpty()) {
+      return planner.getCostFactory().makeCost(rowCount, 0, 0);
+    }
+    final int offsetValue = offset == null ? 0 : RexLiteral.intValue(offset);
+    final double inCount = mq.getRowCount(input);
     // Higher cost if rows are wider discourages pushing a project through a
     // sort.
-    final double rowCount = mq.getRowCount(this);
     final double bytesPerRow = getRowType().getFieldCount() * 4;
-    final double cpu = Util.nLogN(rowCount) * bytesPerRow;
+    // When output count + offset is smaller than input count, we can use heap sort,
+    // otherwise, we can use heap sort, quick sort and so on
+    final double heapSize = Math.min(rowCount + offsetValue, inCount);
```

Comment:
If I am not mistaken, this change could have some side-effects, e.g.  `EnumerableSort` currently does not override `computeSelfCost`, it uses `Sort`'s one.  `EnumerableSort` implements a nLogn sorting algorithm, so the current `Sort#computeSelfCost` suits better than the new proposed one. I think if we modify the formula here, then we should override `EnumerableSort#computeSelfCost` and apply the old formula there.
However, other downstream projects with their own sort operators (which currently rely on the current formula) might be impacted as well, so this could be seen somehow as a breaking change.

Reply:
Query:
select * from foodmart.sales_fact_1997 order by cust_id limit 10
Plan:
EnumerableLimit(fetch=[10]): rowcount = 10.0, cumulative cost = {210.0 rows, 3795.1361487904733 cpu, 0.0 io}, id = 35
  EnumerableSort(sort0=[$0], dir0=[ASC]): rowcount = 100.0, cumulative cost = {200.0 rows, 3785.1361487904733 cpu, 0.0 io}, id = 34
    EnumerableTableScan(table=[[foodmart, sales_fact_1997]]): rowcount = 100.0, cumulative cost = {100.0 rows, 101.0 cpu, 0.0 io}, id = 20

I read the EnumerableSort and found the offset and fetch will be both null, so the sort cost is still nLogN.  code as follow:

  public EnumerableSort(RelOptCluster cluster, RelTraitSet traitSet,
      RelNode input, RelCollation collation, @Nullable RexNode offset, @Nullable RexNode fetch) {
    super(cluster, traitSet, input, collation, offset, fetch);
    assert getConvention() instanceof EnumerableConvention;
    assert getConvention() == input.getConvention();
    assert fetch == null : "fetch must be null";
    assert offset == null : "offset must be null";
  }

Reply:
Yes, your're right, the cost of EnumerableSort will effectively continue being nLogn

Reply:
I found duplicate code in EnumerableLimitSort. Shall I close the pr or change the EnumerableLimitSort. code as follow
```
@Override public @Nullable RelOptCost computeSelfCost(RelOptPlanner planner,
      RelMetadataQuery mq) {
    final double rowCount = mq.getRowCount(this.input).doubleValue();
    double toSort = getValue(this.fetch, rowCount);
    if (this.offset != null) {
      toSort += getValue(this.offset, rowCount);
    }
    // we need to sort at most rowCount rows
    toSort = Math.min(rowCount, toSort);

    // we need to process rowCount rows, and for every row
    // we search the key in a TreeMap with at most toSort entries
    final double lookup = Math.max(1., Math.log(toSort));
    final double bytesPerRow = this.getRowType().getFieldCount() * 4.;
    final double cpu = (rowCount * lookup) * bytesPerRow;

    RelOptCost cost = planner.getCostFactory().makeCost(rowCount, cpu, 0);
    return cost;
  }
```